flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/NetworkInput.java

```diff
@@ -0,0 +1,154 @@
+import org.apache.flink.runtime.io.network.api.serialization.RecordDeserializer.DeserializationResult;
+import org.apache.flink.runtime.io.network.api.serialization.SpillingAdaptiveSpanningRecordDeserializer;
+import org.apache.flink.runtime.io.network.buffer.Buffer;
+import org.apache.flink.runtime.io.network.partition.consumer.BufferOrEvent;
+import org.apache.flink.runtime.io.network.partition.consumer.InputGate;
+import org.apache.flink.runtime.plugable.DeserializationDelegate;
+import org.apache.flink.runtime.plugable.NonReusingDeserializationDelegate;
+import org.apache.flink.streaming.runtime.streamrecord.StreamElement;
+import org.apache.flink.streaming.runtime.streamrecord.StreamElementSerializer;
+
+import java.io.IOException;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+
+/**
+ * Implementation of taking {@link InputGate} as {@link Input}.
+ */
+@Internal
+public final class NetworkInput implements Input {
+
+	private final int numberOfInputChannels;
+
+	private final CheckpointBarrierHandler barrierHandler;
+
+	private final DeserializationDelegate<StreamElement> deserializationDelegate;
+
+	private final RecordDeserializer<DeserializationDelegate<StreamElement>>[] recordDeserializers;
+
+	/**
+	 * The channel from which a buffer came, tracked so that we can appropriately map
+	 * the watermarks and watermark statuses to the correct channel index of the correct valve.
+	 */
+	private int currentChannel = -1;
+
+	private RecordDeserializer<DeserializationDelegate<StreamElement>> currentRecordDeserializer = null;
+
+	private boolean isFinished = false;
+
+	@SuppressWarnings("unchecked")
+	public NetworkInput(
+			CheckpointBarrierHandler barrierHandler,
+			TypeSerializer<?> inputSerializer,
+			IOManager ioManager) {
+		this.barrierHandler = barrierHandler;
+		this.numberOfInputChannels = barrierHandler.getNumberOfInputChannels();
+		this.deserializationDelegate = new NonReusingDeserializationDelegate<>(
+			new StreamElementSerializer<>(inputSerializer));
+
+		// Initialize one deserializer per input channel
+		this.recordDeserializers = new SpillingAdaptiveSpanningRecordDeserializer[numberOfInputChannels];
+		for (int i = 0; i < recordDeserializers.length; i++) {
+			recordDeserializers[i] = new SpillingAdaptiveSpanningRecordDeserializer<>(
+				ioManager.getSpillingDirectoriesPaths());
+		}
+	}
+
+	@Override
+	public StreamElement pollNext() throws Exception {
+
+		while (true) {
+			// get the stream element from the deserializer
+			if (currentRecordDeserializer != null) {
+				DeserializationResult result = currentRecordDeserializer.getNextRecord(deserializationDelegate);
+				if (result.isBufferConsumed()) {
+					currentRecordDeserializer.getCurrentBuffer().recycleBuffer();
+					currentRecordDeserializer = null;
+				}
+
+				if (result.isFullRecord()) {
+					return deserializationDelegate.getInstance();
+				}
+			}
+
+			Optional<BufferOrEvent> bufferOrEvent = barrierHandler.pollNext();
```

Comment:
In `StreamTwoInputSeletableProcessor`, `input1` and `input2` each correspond to a `NetworkInput`, and each `NetworkInput` should contain its own `InputGate` (`inputGate1` or `inputGate2`). But I guess your `BarrierHandler` should contain one `InputGate` unioned `inputGate1` and `inputGate2` (`BarrierHandler`  must see all input channels), so I think the current implementation of `BarrierHandler` may not meet the requirements.

Reply:
Got it.